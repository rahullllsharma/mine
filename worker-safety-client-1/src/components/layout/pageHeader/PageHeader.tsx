import type { PropsWithClassName, RouterLink } from "@/types/Generic";

import type { ReactNode } from "react";
import type {
  PageHeaderAction,
  PageHeaderActionTooltip,
} from "./components/headerActions/HeaderActions";
import cx from "classnames";
import NextLink from "next/link";
import { capitalize } from "lodash-es";
import { FormViewTabStates } from "@/components/forms/Utils";
import Link from "@/components/shared/link/Link";

import { HeaderActions } from "./components/headerActions/HeaderActions";

type PageHeaderBaseProps = PropsWithClassName<{
  children?: ReactNode;
  pageTitle?: string;
  linkText?: string;
  linkRoute?: string | RouterLink;
  actions?: PageHeaderActionTooltip | PageHeaderAction | PageHeaderAction[];
  additionalInfo?: ReactNode;
  selectedTab?: FormViewTabStates;
  setSelectedTab?: (selectedTab: FormViewTabStates) => void;
}>;

type PageHeaderPropsWithLink = PageHeaderBaseProps & {
  linkText: string;
  linkRoute: string | RouterLink;
};

type PageHeaderPropsWithoutLink = PageHeaderBaseProps & {
  linkText?: never;
  linkRoute?: never;
};

type PageHeaderProps = PageHeaderPropsWithLink | PageHeaderPropsWithoutLink;

export default function PageHeader({
  className,
  children,
  pageTitle,
  linkText,
  linkRoute,
  actions,
  additionalInfo,
  setSelectedTab,
  selectedTab,
}: PageHeaderProps): JSX.Element {
  let actionsArray:
    | PageHeaderAction[]
    | [PageHeaderAction]
    | [PageHeaderActionTooltip] = [];

  if (actions) {
    actionsArray = Array.isArray(actions)
      ? actions
      : ([actions] as [PageHeaderAction] | [PageHeaderActionTooltip]);
  }

  const hasActions = actionsArray.length > 0;

  return (
    <header className={cx("shadow-5 py-2 px-6 bg-white relative", className)}>
      {linkRoute && linkText ? (
        <NextLink href={linkRoute} passHref>
          <Link label={linkText} iconLeft="chevron_big_left" className="mb-3" />
        </NextLink>
      ) : (
        <div className="h-5 mb-3" />
      )}
      <div className="flex items-center flex-wrap">
        <div className="flex w-full">
          {pageTitle ? (
            <h4 className="text-neutral-shade-100 mr-3">{pageTitle}</h4>
          ) : (
            children
          )}
          {hasActions && <HeaderActions actions={actionsArray} />}
        </div>
        {additionalInfo && <div className="mt-3 w-full">{additionalInfo}</div>}

        {selectedTab && setSelectedTab && (
          <div className="-mb-2 border-b border-gray-200 dark:border-gray-700 w-full">
            <ul
              className="flex flex-wrap -mb-px text-16px font-medium text-center w-full"
              id="default-tab"
              data-tabs-toggle="#tabs"
              role="tablist"
            >
              {Object.values(FormViewTabStates).map(tab => (
                <li className="me-2" role="presentation" key={tab}>
                  <button
                    className={cx("inline-block p-4 border-b-2 rounded-t-lg", {
                      ["border-brand-urbint-40"]: selectedTab === tab,
                      ["border-transparent hover:text-gray-600 hover:border-gray-300 dark:hover:text-gray-300"]:
                        selectedTab !== tab,
                    })}
                    id={`${tab}-tab`}
                    data-tabs-target={`#${tab}`}
                    type="button"
                    role="tab"
                    aria-controls={tab}
                    aria-selected={selectedTab === tab}
                    onClick={() => setSelectedTab(tab)}
                  >
                    {capitalize(tab)}
                  </button>
                </li>
              ))}
            </ul>
          </div>
        )}
      </div>
    </header>
  );
}

export type { PageHeaderProps };
