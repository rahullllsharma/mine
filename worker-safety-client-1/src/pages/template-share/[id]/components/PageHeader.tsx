import { useRouter } from "next/router";
import { useState } from "react";
import type { MenuItemProps } from "@/components/shared/dropdown/Dropdown";
import axiosRest from "@/api/customFlowApi";
import useRestMutation from "@/hooks/useRestMutation";
import { Subheading, IconName } from "@urbint/silica";
import StatusBadge from "@/components/statusBadge/StatusBadge";
import Dropdown from "@/components/shared/dropdown/Dropdown";
import ButtonIcon from "@/components/shared/button/icon/ButtonIcon";

const PageHeader = ({
  title,
  status
}: {title: string
  status: string
}) => {
  const {
    query: { id },
  } = useRouter();
  const [isDownloading, setIsDownloading] = useState(false); 

  const { mutate: downloadPDF } = useRestMutation<any>({
    endpoint: `/pdf_download/forms/${id}`,
    method: "get",
    axiosInstance: axiosRest,
    axiosConfig: {
      responseType: "blob",
    },
    mutationOptions: {
      onSuccess: async (response: any) => {
        setIsDownloading(false); 

        const blob = new Blob([response.data], { type: "application/pdf" });
        console.log("PDF disposition", response.headers);
        const disposition = response.headers?.["content-disposition"];
        let filename = `form-${Date.now()}.pdf`;
        if (disposition) {
          const match = disposition.match(/filename[^;=\n]*=((['"]).*?\2|[^;\n]*)/);
          if (match && match[1]) {
            filename = match[1].replace(/['"]/g, "");
          }
        }
        const url = window.URL.createObjectURL(blob);
        const link = document.createElement("a");
        link.href = url;
        link.setAttribute("download", filename);
        document.body.appendChild(link);
        link.click();
        link.remove();
        window.URL.revokeObjectURL(url);
      },
      onError: () => {
        setIsDownloading(false); // Stop loader on error
      },
    },
  });

  const handleDownloadPDF = () => {
    setIsDownloading(true);
    downloadPDF({});
  };

  const menuItems: MenuItemProps[] = [
    {
      label: isDownloading ? "Downloading..." : "Download PDF",
      icon: isDownloading ? ("spinner" as IconName) : ("download" as IconName),
      onClick: handleDownloadPDF,
      disabled: isDownloading, 
    },
  ];

  return (
    <div className="p-4 bg-white shadow-5 flex justify-between items-center">
      <div className="flex items-center relative">
        <Subheading className="text-[26px] py-2 m-0 mr-2">
          {title}
        </Subheading>
        <StatusBadge status={status} />
      </div>
      {status === "completed" && (
        <Dropdown
          className="z-10"
          menuItems={[menuItems]}
        >
          <ButtonIcon iconName="hamburger" />
        </Dropdown>
      )}
    </div>
  );
};

export default PageHeader;
