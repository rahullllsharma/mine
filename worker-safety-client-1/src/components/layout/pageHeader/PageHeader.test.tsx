import type {
  PageHeaderAction,
  PageHeaderActionTooltip,
} from "./components/headerActions/HeaderActions";
import { fireEvent, render, screen } from "@testing-library/react";
import PageHeader from "./PageHeader";

describe(PageHeader.name, () => {
  it("should be rendered correctly", () => {
    const { asFragment } = render(
      <PageHeader linkText="some label" linkRoute="route" />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should display the link text and page title", () => {
    const linkLabel = "go back";
    const pageTitle = "tasks page";

    render(
      <PageHeader
        linkText={linkLabel}
        linkRoute="route"
        pageTitle={pageTitle}
      />
    );
    screen.getByText(linkLabel);
    screen.getByText(pageTitle);
  });

  describe("when a tooltip is passed in the header props", () => {
    beforeEach(() => {
      const headerTooltip: PageHeaderActionTooltip = {
        type: "tooltip",
        title: "tooltip hover",
      };

      render(<PageHeader actions={headerTooltip} />);
    });

    it("should display a icon non-interactive", () => {
      screen.getByRole("tooltip", { hidden: true });
    });

    it("when hovered, it should display the message", () => {
      fireEvent.mouseOver(screen.getByRole("tooltip", { hidden: true }));
      expect(screen.getByText("tooltip hover")).toBeInTheDocument();
    });
  });

  describe("when a single header action is passed in the header props", () => {
    it("should display the clickable icon", () => {
      const headerAction: PageHeaderAction = {
        icon: "settings_filled",
        title: "settings",
        onClick: jest.fn(),
      };

      render(<PageHeader actions={headerAction} />);
      screen.getByRole("button", { name: /settings/i });
    });
  });

  describe("when an header actions array with a single element is passed in the header props", () => {
    it("should display the clickable icon", () => {
      const headerAction: PageHeaderAction = {
        icon: "settings_filled",
        title: "settings",
        onClick: jest.fn(),
      };

      render(<PageHeader actions={[headerAction]} />);
      screen.getByRole("button", { name: /settings/i });
    });
  });

  describe("when an header actions array with multiple elements is passed in the header props", () => {
    it("should display a dropdown menu", () => {
      const headerActions: PageHeaderAction[] = [
        {
          icon: "settings_filled",
          title: "settings",
          onClick: jest.fn(),
        },
        {
          icon: "trash_empty",
          title: "delete",
          onClick: jest.fn(),
        },
      ];

      render(<PageHeader actions={headerActions} />);
      screen.getAllByTestId("dropdown");
    });
  });
});
