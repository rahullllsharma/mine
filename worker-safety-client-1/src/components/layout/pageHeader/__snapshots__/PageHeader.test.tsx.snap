// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageHeader should be rendered correctly 1`] = `
<DocumentFragment>
  <header
    class="shadow-5 py-2 px-6 bg-white relative"
  >
    <a
      class="font-medium text-brand-urbint-50 leading-5 cursor-pointer inline-flex items-center active:text-brand-urbint-60 hover:text-brand-urbint-40 mb-3"
      href="/route"
    >
      <i
        aria-hidden="true"
        class="ci-chevron_big_left text-lg leading-5 mr-1"
      />
      <span
        class="truncate"
      >
        some label
      </span>
    </a>
    <div
      class="flex items-center flex-wrap"
    >
      <div
        class="flex w-full"
      />
    </div>
  </header>
</DocumentFragment>
`;
